ArduSub/control_motordetect.cpp

```diff
@@ -0,0 +1,175 @@
+#include "Sub.h"
+#include "stdio.h"
+
+/*
+ * control_motordetect.cpp - init and run calls for motordetect flightmode;
+ *
+ *  This mode pulses all thrusters to detect if they need to be reversed.
+ *  This still requires that the user has the correct frame selected and the motors
+ *  are connected to the correct ESCs.
+ *
+ *  For each motor:
+ *      wait until vehicle is stopped for > 500ms
+ *      apply throttle up for 500ms
+ *      If results are good:
+ *          save direction and try the next motor.
+ *      else
+ *          wait until vehicle is stopped for > 500ms
+ *          apply throttle down for 500ms
+ *          If results are good
+ *              save direction and try the next motor.
+ *          If results are bad
+ *              Abort!
+ */
+
+// controller states
+enum test_state {
+    STANDBY,
+    SETTLING,
+    THRUSTING,
+    DETECTING,
+    DONE
+};
+
+enum direction {
+    UP = 1,
+    DOWN = -1
+};
+
+static uint32_t settling_timer;
+static uint32_t thrusting_timer;
+static uint8_t md_state;
+static uint8_t current_motor;
+static int16_t current_direction;
+
+
+bool Sub::motordetect_init()
+{
+    current_motor = 0;
+    md_state = STANDBY;
+    current_direction = UP;
+    return true;
+}
+
+void Sub::motordetect_run()
+{
+    // if not armed set throttle to zero and exit immediately
+    if (!motors.armed()) {
+        motors.set_desired_spool_state(AP_Motors::DesiredSpoolState::GROUND_IDLE);
+        // Force all motors to stop
+        for(uint8_t i=0; i < AP_MOTORS_MAX_NUM_MOTORS; i++) {
+            if (motors.motor_is_enabled(i)) {
+                motors.output_test_num(i, 1500);
+            }
+        }
+        md_state = DONE;
+        return;
+    }
+
+    switch(md_state) {
+    // Motor detection is not running, set it up to start.
+    case STANDBY:
+        current_direction = UP;
+        current_motor = 0;
+        settling_timer = AP_HAL::millis();
+        md_state = SETTLING;
+        break;
+
+    // Wait until sub stays for 500ms not spinning and leveled.
+    case SETTLING:
+        // Force all motors to stop
+        for (uint8_t i=0; i <AP_MOTORS_MAX_NUM_MOTORS; i++) {
+            if (motors.motor_is_enabled(i)) {
+                !motors.output_test_num(i, 1500);
+            }
+        }
+        // wait until roll and pitch are between -10 and 10 degrees
+        // and gyro product is under a certain(experimental) threshold
+        if (abs(ahrs.roll_sensor) > 1000
```

Comment:
I think we can omit this 'wait for level' step. Static stability should not be a requirement.

Reply:
you are right, it is gone now